STORM-2497: Let Supervisor enforce memory and add in support for shared memory regions

revans2 opened the pull request:
This is based off of #2112 so that the formatting is simpler.

jerrypeng commented:
cool feature

revans2 commented:
@kishorvpatil Could you look again.  I had to rebase becase of a minor conflict in the DefaultResourceAwareStrategy.  It ended up resulting in only some comment changes to this code, but I wanted to be sure you took a look.

srdo commented on storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -728,8 +758,145 @@ protected String javaCmd(String cmd) {
+        typeOfCheck = "WORKER " + _workerId;
+      }
+      LOG.debug(
+          "Enforcing memory usage for {} with usgae of {} out of {} total and a hard limit of {}",
```
usgae -> usage

srdo commented on storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -728,8 +758,145 @@ protected String javaCmd(String cmd) {
+          hardMemoryLimitOver);
+
+      if (usageMb <= 0) {
+        //Looks like usage might now be supported
```
now -> not(?)

srdo commented on storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -728,8 +758,145 @@ protected String javaCmd(String cmd) {
+      long hardLimitMb =
+          memoryLimitMb
+              + Math.max(
+                  (long) (memoryLimitMb * (hardMemoryLimitMultiplier - 1.0)), hardMemoryLimitOver);
```
Nit: I think this would be easier to read as max(memoryLimit*hardMultiplier, memoryLimit + hardLimitOver), rather than adding first and then adjusting the multiplier.

srdo commented on storm-server/src/main/java/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -528,28 +577,108 @@ public void cleanUpForRestart() throws IOException {
+     * that it just went over the limit.
+     * @throws IOException on any error
+     */
+    public boolean isMemoryLimitViolated(LocalAssignment withUpdatedLimits) throws IOException {
```
This doesn't seem like it can throw IOException

revans2 replied:
Yes but it is overridden by other Container Implementations and they may want to throw an exception.

srdo commented on storm-server/src/main/java/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -559,9 +688,17 @@ public void cleanUpForRestart() throws IOException {
+    /**
+     * Get the topology id this is a part of.
+     */
+    public String getTopoogyId() {
```
topoogy -> topology

srdo commented on storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java:
```diff
@@ -0,0 +1,49 @@
+import org.junit.Test;
+
+public class NimbusTest {
+    @Test(expected=IllegalArgumentException.class)
```
Nit: Could reduce the scope of this assertion more local with the ExpectedException rule

srdo commented on storm-server/src/test/java/org/apache/storm/scheduler/resource/TestResourceAwareScheduler.java:
```diff
@@ -46,53 +44,43 @@
-        defaultTopologyConf.put(Config.TOPOLOGY_COMPONENT_CPU_PCORE_PERCENT, 10.0);
-        defaultTopologyConf.put(Config.TOPOLOGY_COMPONENT_RESOURCES_ONHEAP_MEMORY_MB, 128.0);
-        defaultTopologyConf.put(Config.TOPOLOGY_COMPONENT_RESOURCES_OFFHEAP_MEMORY_MB, 0.0);
+      //TODO clean this up some more
```
Todo

srdo commented on storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -144,8 +148,19 @@ public Void call() throws Exception {
+                    Map<String, Object> topoConf = ConfigUtils.readSupervisorStormConf(_conf, _topologyId);
+                    _fsOps.setupStormCodeDir(topoConf, _stormRoot);
+                    if (_assignment.is_has_node_shared_memory()) {
+                        File sharedMemoryDir = new File(_stormRoot, "shared_by_topology");
```
I might have missed it somewhere, but how does the topology code access the off heap shared memory?

revans2 replied:
That is up to the topology.  It could be JNI code or even something that a ShellBolt or ShellSpout does.  The point is to provide a place for them to store the data if they need it.  We have a group doing this in /tmp right now and causing all kinds of issues if things crash/get rescheduled leaking things.

srdo replied:
Okay. I was just thinking it might be helpful if the topology could get the directory path (or _stormRoot), so users don't have to figure out how to path to this directory themselves.

revans2 replied:
I can add it to the documentation.  But I didn't spend a lot of time documenting it on purpose.  shared memory off heap is hard to get right and I would rather discourage people from using it.

srdo commented on storm-server/src/main/java/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -151,22 +161,40 @@ public void reserveResourcesForWorker(String workerId, Map<String, Number> resou
+                    (int)
+                        (Math.ceil(
+                            ObjectReader.getDouble(
+                                this.conf.get(DaemonConfig.STORM_CGROUP_MEMORY_LIMIT_TOLERANCE_MARGIN_MB),
```
I'm wondering if having defaults here is necessary? Isn't defaults.yaml always expected to be present?

revans2 replied:
Except in some unit tests.

srdo replied:
Makes sense

revans2 commented on storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignmentImpl.java:
```diff
@@ -19,36 +19,82 @@
+    public SchedulerAssignmentImpl(String topologyId, Map<ExecutorDetails, WorkerSlot> executorToSlot,
+            Map<WorkerSlot, WorkerResources> resources, Map<String, Double> totalSharedOffHeap) {
+        this.topologyId = topologyId;       
+        if (executorToSlot != null) {
```
I prefer to be very lenient with inputs if something can be null I don't want to force others to never use null.

srdo replied:
Ok. I'm the opposite way, I prefer avoiding null checks. But this is fine :)

revans2 commented on storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignmentImpl.java:
```diff
@@ -141,4 +222,42 @@ public String getTopologyId() {
+    }
+
+    /**
+     * Update the resources for this assignment (This should go aware when the RAS-MT bridge goes away
```
Sorry code I forgot to remove that is not needed here.

revans2 commented on storm-client/src/jvm/org/apache/storm/topology/BaseConfigurationDeclarer.java:
```diff
@@ -52,18 +52,20 @@ public T setNumTasks(Number val) {
+    @SuppressWarnings("unchecked")
     @Override
     public T setMemoryLoad(Number onHeap) {
         if (onHeap != null) {
```
I agree but that is code that came before, and if we want to make that change I would prefer to do it on a separate pull request, just so a breaking change does not sneak in as part of review comments.

revans2 replied:
I filed https://issues.apache.org/jira/browse/STORM-2545 for it

srdo replied:
Thanks

revans2 commented on storm-client/src/jvm/org/apache/storm/trident/operation/DefaultResourceDeclarer.java:
```diff
@@ -29,8 +33,8 @@
 
-    private Map<String, Number> resources = new HashMap<>();
-    private static Map<String, Object> conf = Utils.readStormConfig();
+    private final transient Map<String, Number> resources = new HashMap<>();
```
In trident Node inherits from DefaultResourceDeclarer, and Node is serialized out (repeatedly for different bolts and spouts in the topology).  These are never used after the topology is created so this reduces the serialized size of a trident topology.

I'll add a comment in the code to make it clean why.

revans2 commented:
I think I addressed all of the review comments

srdo commented on storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignment.java:
```diff
@@ -21,41 +21,52 @@
     /**
-     * get the topology-id this assignment is for.
-     * @return
+     * Return the ID of the topolgoy.
```
topolgoy

revans2 replied:
I even typo my fixes.

srdo commented on storm-client/src/jvm/org/apache/storm/trident/operation/DefaultResourceDeclarer.java:
```diff
@@ -29,8 +34,12 @@
 
-    private Map<String, Number> resources = new HashMap<>();
-    private static Map<String, Object> conf = Utils.readStormConfig();
+    //@{link org.apache.storm.trident.planner.Node} and several other tirdent classes inherit from DefaultResourceDeclarer
```
tirdent

srdo commented on storm-client/src/jvm/org/apache/storm/trident/operation/DefaultResourceDeclarer.java:
```diff
@@ -29,8 +34,12 @@
-    private static Map<String, Object> conf = Utils.readStormConfig();
+    //@{link org.apache.storm.trident.planner.Node} and several other tirdent classes inherit from DefaultResourceDeclarer
+    // These classes are serialized out as part of the bolts and spouts of a topology, often for each bolt/spout in the topology.
+    // The following are marked as transiant because they are never used after the topology is created so keeping them around just wasts
```
wasts

srdo commented on storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java:
```diff
@@ -24,8 +24,10 @@
 import org.apache.storm.topology.TopologyBuilder;
 import org.junit.Test;
 
+import static org.junit.Assert.*;
```
I think checkstyle bans this

revans2 replied:
Checkstyle does not check tests for some reason, but if you care about it I can change it.

srdo replied:
No, don't worry about it. Hadn't noticed this was in a test

srdo commented on storm-server/src/test/java/org/apache/storm/daemon/nimbus/NimbusTest.java:
```diff
@@ -44,6 +46,11 @@ public void testMemoryLoadLargerThanMaxHeapSize() throws Exception {
-        Nimbus.validateTopologyWorkerMaxHeapSizeConfigs(config1, stormTopology1);
+        try {
+            Nimbus.validateTopologyWorkerMaxHeapSizeConfigs(config1, stormTopology1);
+            fail("Expected exeption not thrown");
```
exeption

revans2 commented:
@srdo I think I got the rest of them

@kishorvpatil could you take a look at let me know if the rework is good for you?

revans2 commented:
@srdo @kishorvpatil @jerrypeng 

I just rebased this on the latest master and added in a few caches in some of the internal data structures to help with performance issues we had seen in production.  Please take a look again, I really would like to get this in sooner rather then later as we have some other work we want to do on the RAS scheduler to add in generic resource support (we have a goal to officially support GPUs)

srdo commented:
+1, though I'll admit to just skimming it this time. It would be easier to tell what changed if the non-merge-conflict-resolution changes were in separate commits.

It might make sense to drop a line to @roshannaik before merging, I think he asked on the mailing list to be made aware of changes to storm-client due to https://github.com/apache/storm/pull/2241
